R/cross_classify_helpers.R

```diff
@@ -0,0 +1,136 @@
+                            result))
+
+  # Cleanup: remove row names
+  rownames(merged_data) = NULL
+  # Re-write the column names to be the original column names from the original
+  # dfs.
+  colnames(merged_data) = unname(unlist(lapply(dfs, colnames)))
+
+  merged_data
+}
+
+joint_draw_ecdf = function (data_list, N, ndim=length(data_list),
+                            sigma=NULL, rho=0, use_f = TRUE) {
+
+  # We don't modify data_list, but this is useful to ensure the
+  # argument is evaluated anyway
+  force(ndim)
+
+  # Error handling for N
+  if(is.null(N) || is.na(N) || !is.atomic(N) || length(N) > 1 || N <= 0) {
+    stop("N must be a single integer that is positive.")
+  }
+
+  # Error handling for rho, if specified
+  if(is.null(sigma)) {
+    if(is.atomic(rho) & length(rho)==1) {
+      if(ndim>2 & rho<0) {
+        stop("The correlation matrix must be positive semi-definite. In specific, ",
+             "if the number of variables being drawn from jointly is 3 or more, ",
+             "then the correlation coefficient rho must be non-negative.")
+      }
+
+      if(rho == 0) {
+        # Uncorrelated draw would be way faster; just sample each column
+        return(lapply(seq_along(data_list),
+                                       function(vn) {
+                                         sample.int(length(data_list[[vn]]), N, replace=TRUE)
+                                       }))
+      }
+      sigma = matrix(rho, ncol=ndim, nrow=ndim)
+      diag(sigma) = 1
+    } else {
+      stop("If specified, rho should be a single number")
+    }
+  }
+
+  # Error handling for sigma
+  if(ncol(sigma) != ndim | nrow(sigma) != ndim | any(diag(sigma) != 1)) {
+    stop("The correlation matrix must be square, with the number of dimensions ",
+         "equal to the number of dimensions you are drawing from. In addition, ",
+         "the diagonal of the matrix must be equal to 1.")
+  }
+
+  # Can we use the fast package or are we stuck with the slow one?
+  use_f = use_f && requireNamespace("mvnfast", quietly = TRUE)
+
+  # Standard normal = all dimensions mean 0.
+  mu = rep(0, ndim)
+
+  # Possible options for the joint normal draw
+  if(!use_f) {
+    # Below code is a reimplementation of the operative parts of rmvnorm from
+    # the mvtnorm package so that we don't induce a dependency
+
+    # Right cholesky decomposition (i.e. LR = sigma s.t. L is lower triang, R
+    # is upper triang.)
+    right_chol = chol(sigma, pivot=TRUE)
+    # Order columns by the pivot attribute -- induces numerical stability?
+    right_chol = right_chol[, order(attr(right_chol, "pivot"))]
+    # Generate standard normal data and right-multiply by decomposed matrix
+    # with right_chol to make it correlated.
+    correlated_sn <- matrix(rnorm(N * ndim),
+                            nrow = N,
+                            byrow = TRUE) %*% right_chol
+
+  } else {
+    # Using mvnfast
+    correlated_sn = mvnfast::rmvn(N, ncores = 2, mu, sigma)
```

Comment:
`getOption("mc.cores")`

Reply:
oh, nuts, this is clearly an artifact from putzing around on my machine. Will fix.